File: src/backend/pcp/src/model/pcp_servico_model.py
from src import db
from src.model.pcp_servico_colaborador_model import PcpServicoColaboradorModel
from src.model.pcp_servico_equipamento_model import PcpServicoEquipamentoModel
from src.model.pcp_op_detalhe_model import PcpOpDetalheModel


class PcpServicoModel(db.Model):
    __tablename__ = 'pcp_servico'

    id = db.Column(db.Integer, primary_key=True)
    inicio_realizado = db.Column(db.DateTime)
    termino_realizado = db.Column(db.DateTime)
    horas_realizado = db.Column(db.Integer)
    minutos_realizado = db.Column(db.Integer)
    segundos_realizado = db.Column(db.Integer)
    custo_realizado = db.Column(db.Float)
    inicio_previsto = db.Column(db.DateTime)
    termino_previsto = db.Column(db.DateTime)
    horas_previsto = db.Column(db.Integer)
    minutos_previsto = db.Column(db.Integer)
    segundos_previsto = db.Column(db.Integer)
    custo_previsto = db.Column(db.Float)
    id_pcp_op_detalhe = db.Column(db.Integer, db.ForeignKey('pcp_op_detalhe.id'))

    pcp_servico_colaborador_model_list = db.relationship('PcpServicoColaboradorModel', lazy='dynamic')
    pcp_servico_equipamento_model_list = db.relationship('PcpServicoEquipamentoModel', lazy='dynamic')
    pcp_op_detalhe_model = db.relationship('PcpOpDetalheModel', foreign_keys=[id_pcp_op_detalhe])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_pcp_op_detalhe = data.get('idPcpOpDetalhe')
        self.inicio_realizado = data.get('inicioRealizado')
        self.termino_realizado = data.get('terminoRealizado')
        self.horas_realizado = data.get('horasRealizado')
        self.minutos_realizado = data.get('minutosRealizado')
        self.segundos_realizado = data.get('segundosRealizado')
        self.custo_realizado = data.get('custoRealizado')
        self.inicio_previsto = data.get('inicioPrevisto')
        self.termino_previsto = data.get('terminoPrevisto')
        self.horas_previsto = data.get('horasPrevisto')
        self.minutos_previsto = data.get('minutosPrevisto')
        self.segundos_previsto = data.get('segundosPrevisto')
        self.custo_previsto = data.get('custoPrevisto')

    def serialize(self):
        return {
            'id': self.id,
            'idPcpOpDetalhe': self.id_pcp_op_detalhe,
            'inicioRealizado': self.inicio_realizado.isoformat(),
            'terminoRealizado': self.termino_realizado.isoformat(),
            'horasRealizado': self.horas_realizado,
            'minutosRealizado': self.minutos_realizado,
            'segundosRealizado': self.segundos_realizado,
            'custoRealizado': self.custo_realizado,
            'inicioPrevisto': self.inicio_previsto.isoformat(),
            'terminoPrevisto': self.termino_previsto.isoformat(),
            'horasPrevisto': self.horas_previsto,
            'minutosPrevisto': self.minutos_previsto,
            'segundosPrevisto': self.segundos_previsto,
            'custoPrevisto': self.custo_previsto,
            'pcpServicoColaboradorModelList': [pcp_servico_colaborador_model.serialize() for pcp_servico_colaborador_model in self.pcp_servico_colaborador_model_list],
            'pcpServicoEquipamentoModelList': [pcp_servico_equipamento_model.serialize() for pcp_servico_equipamento_model in self.pcp_servico_equipamento_model_list],
            'pcpOpDetalheModel': self.pcp_op_detalhe_model.serialize() if self.pcp_op_detalhe_model else None,
        }

File: frontend/pcp/lib/app/data/model/pcp_servico_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:pcp/app/infra/infra_imports.dart';
import 'package:pcp/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class PcpServicoModel {
	int? id;
	int? idPcpOpDetalhe;
	DateTime? inicioRealizado;
	DateTime? terminoRealizado;
	int? horasRealizado;
	int? minutosRealizado;
	int? segundosRealizado;
	double? custoRealizado;
	DateTime? inicioPrevisto;
	DateTime? terminoPrevisto;
	int? horasPrevisto;
	int? minutosPrevisto;
	int? segundosPrevisto;
	double? custoPrevisto;
	List<PcpServicoColaboradorModel>? pcpServicoColaboradorModelList;
	List<PcpServicoEquipamentoModel>? pcpServicoEquipamentoModelList;
	PcpOpDetalheModel? pcpOpDetalheModel;

	PcpServicoModel({
		this.id,
		this.idPcpOpDetalhe,
		this.inicioRealizado,
		this.terminoRealizado,
		this.horasRealizado,
		this.minutosRealizado,
		this.segundosRealizado,
		this.custoRealizado,
		this.inicioPrevisto,
		this.terminoPrevisto,
		this.horasPrevisto,
		this.minutosPrevisto,
		this.segundosPrevisto,
		this.custoPrevisto,
		this.pcpServicoColaboradorModelList,
		this.pcpServicoEquipamentoModelList,
		this.pcpOpDetalheModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'inicio_realizado',
		'termino_realizado',
		'horas_realizado',
		'minutos_realizado',
		'segundos_realizado',
		'custo_realizado',
		'inicio_previsto',
		'termino_previsto',
		'horas_previsto',
		'minutos_previsto',
		'segundos_previsto',
		'custo_previsto',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Inicio Realizado',
		'Termino Realizado',
		'Horas Realizado',
		'Minutos Realizado',
		'Segundos Realizado',
		'Custo Realizado',
		'Inicio Previsto',
		'Termino Previsto',
		'Horas Previsto',
		'Minutos Previsto',
		'Segundos Previsto',
		'Custo Previsto',
	];

	PcpServicoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idPcpOpDetalhe = jsonData['idPcpOpDetalhe'];
		inicioRealizado = jsonData['inicioRealizado'] != null ? DateTime.tryParse(jsonData['inicioRealizado']) : null;
		terminoRealizado = jsonData['terminoRealizado'] != null ? DateTime.tryParse(jsonData['terminoRealizado']) : null;
		horasRealizado = jsonData['horasRealizado'];
		minutosRealizado = jsonData['minutosRealizado'];
		segundosRealizado = jsonData['segundosRealizado'];
		custoRealizado = jsonData['custoRealizado']?.toDouble();
		inicioPrevisto = jsonData['inicioPrevisto'] != null ? DateTime.tryParse(jsonData['inicioPrevisto']) : null;
		terminoPrevisto = jsonData['terminoPrevisto'] != null ? DateTime.tryParse(jsonData['terminoPrevisto']) : null;
		horasPrevisto = jsonData['horasPrevisto'];
		minutosPrevisto = jsonData['minutosPrevisto'];
		segundosPrevisto = jsonData['segundosPrevisto'];
		custoPrevisto = jsonData['custoPrevisto']?.toDouble();
		pcpServicoColaboradorModelList = (jsonData['pcpServicoColaboradorModelList'] as Iterable?)?.map((m) => PcpServicoColaboradorModel.fromJson(m)).toList() ?? [];
		pcpServicoEquipamentoModelList = (jsonData['pcpServicoEquipamentoModelList'] as Iterable?)?.map((m) => PcpServicoEquipamentoModel.fromJson(m)).toList() ?? [];
		pcpOpDetalheModel = jsonData['pcpOpDetalheModel'] == null ? PcpOpDetalheModel() : PcpOpDetalheModel.fromJson(jsonData['pcpOpDetalheModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idPcpOpDetalhe'] = idPcpOpDetalhe != 0 ? idPcpOpDetalhe : null;
		jsonData['inicioRealizado'] = inicioRealizado != null ? DateFormat('yyyy-MM-ddT00:00:00').format(inicioRealizado!) : null;
		jsonData['terminoRealizado'] = terminoRealizado != null ? DateFormat('yyyy-MM-ddT00:00:00').format(terminoRealizado!) : null;
		jsonData['horasRealizado'] = horasRealizado;
		jsonData['minutosRealizado'] = minutosRealizado;
		jsonData['segundosRealizado'] = segundosRealizado;
		jsonData['custoRealizado'] = custoRealizado;
		jsonData['inicioPrevisto'] = inicioPrevisto != null ? DateFormat('yyyy-MM-ddT00:00:00').format(inicioPrevisto!) : null;
		jsonData['terminoPrevisto'] = terminoPrevisto != null ? DateFormat('yyyy-MM-ddT00:00:00').format(terminoPrevisto!) : null;
		jsonData['horasPrevisto'] = horasPrevisto;
		jsonData['minutosPrevisto'] = minutosPrevisto;
		jsonData['segundosPrevisto'] = segundosPrevisto;
		jsonData['custoPrevisto'] = custoPrevisto;
		
		var pcpServicoColaboradorModelLocalList = []; 
		for (PcpServicoColaboradorModel object in pcpServicoColaboradorModelList ?? []) { 
			pcpServicoColaboradorModelLocalList.add(object.toJson); 
		}
		jsonData['pcpServicoColaboradorModelList'] = pcpServicoColaboradorModelLocalList;
		
		var pcpServicoEquipamentoModelLocalList = []; 
		for (PcpServicoEquipamentoModel object in pcpServicoEquipamentoModelList ?? []) { 
			pcpServicoEquipamentoModelLocalList.add(object.toJson); 
		}
		jsonData['pcpServicoEquipamentoModelList'] = pcpServicoEquipamentoModelLocalList;
		jsonData['pcpOpDetalheModel'] = pcpOpDetalheModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idPcpOpDetalhe = plutoRow.cells['idPcpOpDetalhe']?.value;
		inicioRealizado = Util.stringToDate(plutoRow.cells['inicioRealizado']?.value);
		terminoRealizado = Util.stringToDate(plutoRow.cells['terminoRealizado']?.value);
		horasRealizado = plutoRow.cells['horasRealizado']?.value;
		minutosRealizado = plutoRow.cells['minutosRealizado']?.value;
		segundosRealizado = plutoRow.cells['segundosRealizado']?.value;
		custoRealizado = plutoRow.cells['custoRealizado']?.value?.toDouble();
		inicioPrevisto = Util.stringToDate(plutoRow.cells['inicioPrevisto']?.value);
		terminoPrevisto = Util.stringToDate(plutoRow.cells['terminoPrevisto']?.value);
		horasPrevisto = plutoRow.cells['horasPrevisto']?.value;
		minutosPrevisto = plutoRow.cells['minutosPrevisto']?.value;
		segundosPrevisto = plutoRow.cells['segundosPrevisto']?.value;
		custoPrevisto = plutoRow.cells['custoPrevisto']?.value?.toDouble();
		pcpServicoColaboradorModelList = [];
		pcpServicoEquipamentoModelList = [];
		pcpOpDetalheModel = PcpOpDetalheModel();
		pcpOpDetalheModel?.id = plutoRow.cells['pcpOpDetalheModel']?.value;
	}	

	PcpServicoModel clone() {
		return PcpServicoModel(
			id: id,
			idPcpOpDetalhe: idPcpOpDetalhe,
			inicioRealizado: inicioRealizado,
			terminoRealizado: terminoRealizado,
			horasRealizado: horasRealizado,
			minutosRealizado: minutosRealizado,
			segundosRealizado: segundosRealizado,
			custoRealizado: custoRealizado,
			inicioPrevisto: inicioPrevisto,
			terminoPrevisto: terminoPrevisto,
			horasPrevisto: horasPrevisto,
			minutosPrevisto: minutosPrevisto,
			segundosPrevisto: segundosPrevisto,
			custoPrevisto: custoPrevisto,
			pcpServicoColaboradorModelList: pcpServicoColaboradorModelListClone(pcpServicoColaboradorModelList!),
			pcpServicoEquipamentoModelList: pcpServicoEquipamentoModelListClone(pcpServicoEquipamentoModelList!),
		);			
	}

	pcpServicoColaboradorModelListClone(List<PcpServicoColaboradorModel> pcpServicoColaboradorModelList) { 
		List<PcpServicoColaboradorModel> resultList = [];
		for (var pcpServicoColaboradorModel in pcpServicoColaboradorModelList) {
			resultList.add(
				PcpServicoColaboradorModel(
					id: pcpServicoColaboradorModel.id,
					idColaborador: pcpServicoColaboradorModel.idColaborador,
					idPcpServico: pcpServicoColaboradorModel.idPcpServico,
				)
			);
		}
		return resultList;
	}

	pcpServicoEquipamentoModelListClone(List<PcpServicoEquipamentoModel> pcpServicoEquipamentoModelList) { 
		List<PcpServicoEquipamentoModel> resultList = [];
		for (var pcpServicoEquipamentoModel in pcpServicoEquipamentoModelList) {
			resultList.add(
				PcpServicoEquipamentoModel(
					id: pcpServicoEquipamentoModel.id,
					idPcpServico: pcpServicoEquipamentoModel.idPcpServico,
					idPatrimBem: pcpServicoEquipamentoModel.idPatrimBem,
				)
			);
		}
		return resultList;
	}

	
}